Storage/SQL/SQLiteBookmarksModel.swift

```diff
@@ -780,6 +784,36 @@ open class UnsyncedBookmarksFallbackModelFactory: BookmarksModelFactory {
         self.bufferFactory = SQLiteBookmarksModelFactory(bookmarks: bookmarks, direction: .buffer)
     }
 
+    // This is a special-case class, so here's the special-case behavior to
+    // know how to handle a folder that contains items drawn from different
+    // parts of the database. We look for the special kinds of folders we
+    // nest at the top level, and then we pick a folder to match.
+    public func factoryForIndex(_ index: Int, inFolder folder: BookmarkFolder) -> BookmarksModelFactory {
+        let concatenated: ConcatenatedBookmarkFolder
+        let i: Int
+
+        // We have either just remote and local mobile bookmarks, or we have Desktop Bookmarks
+        // followed by remote and local mobile bookmarks. Handle either.
+        if let prepended = folder as? PrependedBookmarkFolder {
+            guard let c = prepended.main as? ConcatenatedBookmarkFolder else {
```

Comment:
I've opted to do something like this, but to return `self` instead. The only purpose of this method is to special-case the 'left' and 'right' sides of a particular nested concatenated folder, and otherwise to fall through.

(We _could_ make this even more general, and poke at the contents of `ConcatenatedBookmarkFolder.main` and `.append` to find out which side to use, but I figure that's probably a step too far.)